electrum/gui/qml/components/WalletMainView.qml

```diff
@@ -34,21 +33,33 @@ Item {
     }
 
     function openSendDialog() {
-        _sendDialog = sendDialog.createObject(mainView, {invoiceParser: invoiceParser})
-        _sendDialog.open()
-    }
-
-    function closeSendDialog() {
-        if (_sendDialog) {
-            _sendDialog.doClose()
-            _sendDialog = null
-        }
+        var scanner = app.scanDialog.createObject(mainView, {
+            hint: qsTr('Scan an Invoice, an Address, an LNURL-pay, a PSBT or a Channel backup'),
+        })
+        scanner.onFound.connect(function() {
+            var data = scanner.scanData
+            data = data.trim()
+            if (bitcoin.isRawTx(data)) {
+                app.stack.push(Qt.resolvedUrl('TxDetails.qml'), { rawtx: data })
+            } else if (Daemon.currentWallet.isValidChannelBackup(data)) {
+                var dialog = app.messageDialog.createObject(app, {
+                    title: qsTr('Import Channel backup?'),
+                    yesno: true
+                })
+                dialog.accepted.connect(function() {
+                    Daemon.currentWallet.importChannelBackup(data)
+                })
+                dialog.open()
+            } else {
+                invoiceParser.recipient = data
+            }
+            //scanner.destroy()  // TODO
```

Comment:
Do we need an explicit destroy somewhere to avoid a memleak?
I see most dialogs have `onClosed: destroy()`, but not sure how to mimic that.
Also see `main.qml::scanDialog` (line 393).

Reply:
`onClosed` is a signal handler for `Popup`. Since you use a bare `QObject` as the base for `QEQRScanner`, you don't have the plumbing provided by `Popup`, and you need to detect when the activity is closed and emit a signal, which can then be used to destroy the instance in QML.

However, you don't really need to wrap the scanner in a Qt object and manage the lifecycle. Since there's at most 1 scanner active, you can basically also spawn the Activity from a `@pyqtSlot` decorated method in e.g. `AppController`, and just emit signals from `AppController` depending on the result. I think the `Activity` will be auto-deleted by the JVM when it's closed, and you don't need to manage any lifecycle on the QML side.

Reply:
The last part would be something like this:

```
Button {
    onClicked: AppController.showQRScanner()
}

Connections {
    target: AppController
    function onQRScanned(data) {
        // handle data
    }
}
```

Reply:
> you can basically also spawn the Activity from a @pyqtSlot decorated method in e.g. AppController, and just emit signals from AppController depending on the result

Yes I already had something like that a refactor prior, e.g. see
https://github.com/SomberNight/electrum/commit/bc5f5963e249152b11909112d61a0bc9f89b00df
I've changed to `QEQRScanner(QObject)` instead as this way I can easily replace both the `ScanDialog` and the `SendDialog` use cases. In particular, `QEQRScanner` now has an API that is a drop-in replacement for the `main.qml::scanDialog` usages. The AppController signal approach worked nicely for the SendDialog, but I found it not easy to adapt to the ScanDialog usages.

> you don't have the plumbing provided by Popup, and you need to detect when the activity is closed and emit a signal, which can then be used to destroy the instance from QML.
However, you don't really need to wrap the scanner in a Qt object and manage the lifecycle

But just to confirm, if we do use this `QEQRScanner(QObject)` approach, do we need to explicitly call destroy for the object? (/manually manage the lifecycle)
I guess I could just call destroy wherever the code on master calls close/closeDialog.